grafana/resource_folder_permission.go

```diff
@@ -0,0 +1,164 @@
+							Optional: true,
+							Default:  -1,
+						},
+						"user_id": {
+							Type:     schema.TypeInt,
+							Optional: true,
+							Default:  -1,
+						},
+						"permission": {
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: validation.StringInSlice([]string{"View", "Edit", "Admin"}, false),
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func UpdateFolderPermissions(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gapi.Client)
+
+	v, ok := d.GetOk("permissions")
+	if !ok {
+		return nil
+	}
+	permissionList := gapi.PermissionItems{}
+	for _, permission := range v.(*schema.Set).List() {
+		permission := permission.(map[string]interface{})
+		permissionItem := gapi.PermissionItem{}
+		if permission["role"].(string) != "" {
+			permissionItem.Role = permission["role"].(string)
+		}
+		if permission["team_id"].(int) != -1 {
+			permissionItem.TeamId = int64(permission["team_id"].(int))
+		}
+		if permission["user_id"].(int) != -1 {
+			permissionItem.UserId = int64(permission["user_id"].(int))
+		}
+		permissionItem.Permission = mapPermissionStringToInt64(permission["permission"].(string))
+		permissionList.Items = append(permissionList.Items, &permissionItem)
+	}
+
+	folderUID := d.Get("folder_uid").(string)
+
+	err := client.UpdateFolderPermissions(folderUID, &permissionList)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(folderUID)
+
+	return ReadFolderPermissions(d, meta)
+}
+
+func ReadFolderPermissions(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gapi.Client)
+
+	folderUID := d.Get("folder_uid").(string)
+
+	folderPermissions, err := client.FolderPermissions(folderUID)
+	if err != nil {
+		if strings.HasPrefix(err.Error(), "status: 404") {
+			log.Printf("[WARN] removing folder %s from state because it no longer exists in grafana", folderUID)
+			d.SetId("")
+			return nil
+		}
+
+		return err
+	}
+
+	permissionItems := make([]interface{}, len(folderPermissions))
+	count := 0
+	for _, permission := range folderPermissions {
+		if permission.FolderUid != "" {
+			permissionItem := make(map[string]interface{})
+			permissionItem["role"] = permission.Role
+			permissionItem["team_id"] = permission.TeamId
+			permissionItem["user_id"] = permission.UserId
+			permissionItem["permission"] = mapPermissionInt64ToString(permission.Permission)
+
+			permissionItems[count] = permission
+			count++
+		}
+	}
+
+	d.Set("permissions", permissionItems)
+
+	return nil
+}
+
+func DeleteFolderPermissions(d *schema.ResourceData, meta interface{}) error {
+	//there is no delete call for folder permissions. we *could* try to delete
+	//all of the individual permissions for the folder, but that would just leave
+	//us with an inoperable folder. however, in cases where a folder is deleted,
+	//we want the corresponding folder permissions to be removed from state without
+	//making a call to the API which would just fail if the folder is gone.
```

Comment:
I think we do want to remove all permissions. Admins will still be able to access the folder.

![image](https://user-images.githubusercontent.com/908409/96664023-24563a00-130f-11eb-86ce-1228ca87caab.png)

We'd want to check if the folder exists first. If it's gone, no need to make the API call.

Reply:
Yeah this is a good point, but checking if the folder exists is a little more difficult than it should be. The Get/Put for folder permissions are done using the folder uid so that's what I used to tie the permissions to the folder. While the Grafana folder API provides the option to select a folder by UID, the go api does not. So I can switch the permissions to use the folder id and make the uid a computed value, I can make the delete a little more hacky trying to find the folder by uid by selecting all folders and finding the right one, or I can go update the go api to support that extra select by uid call. Any preferences?

Reply:
Since the permissions *should* be referencing the folder.folder_uid wouldn't all of the destroys happen in reverse order? So maybe it's ok to remove permissions without checking for the existence? If someone does something really hacky and hard-codes a uid value and the proper destroy order doesn't happen maybe we can ignore a failure because the parent folder doesn't exist?